Secret life of cows

killercup opened the pull request:
- [x] write initial draft
- [x] links to std docs, e.g. ToOwned and std::collections
- [x] https://llogiq.github.io/2015/07/10/cow-redux.html
- [x] section about examples of cows in the wild
- [x] serde zero-copy deserialization magic

killercup committed:
Draft: Cow Life

killercup committed:
Draft: Cow Life

killercup committed:
Draft: Cow Life

killercup committed:
Draft: Cow Life

mbrubeck commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,129 @@
+---
+
+A lot of people at RustFest Paris mentioned Cows.
+But still, the bovine super powers of Rust's standard library [appaer][1] to be a well-kept secret.
```
typo: "appaer"

mbrubeck commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,129 @@
+Ensuring all these properties is a great way to write fast programs.
+Let's look at how we can do this in Rust:
+
+If you know what you will do with you data,
```
typo: "with you"

mbrubeck commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,129 @@
+
+[1]: https://twitter.com/KevinHoffman/status/1001075501358776322
+
+## Organizing data
```
I like this overview section. It's very clear.

mbrubeck commented:
For "examples of Cows in the wild", here's a case that I thought was interesting:

https://github.com/TeXitoi/benchmarksgame-rs/blob/f78f21bffc68cb42dd9311694913ea798535e674/src/regex_redux.rs#L72-L79

This uses Cow not for getting data into our out of a function, but just so that the same variable can store either borrowed or owned data on different loop iterations.

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,129 @@
+This is what it all comes down to:
+People want to have a good, precise way to organize their data.
+And they want their programming language to support them.
+That's why a lot of newer languages include a bunch of data structures
```
I think the message of this sentence could be improved a little because it's not entirely clear how it connects to "People want to have a good, precise way to organize their data."

killercup committed:
Cow: fix typos

killercup committed:
Cow: reorder refs

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,129 @@
+It's essential to know which piece of data is represented in which way.
+
+In systems programming languages,
+this is in some regards even more important:
```
Maybe add a period at the end, followed by "You want to know:" to start the list.

killercup committed:
Cow: Normalize headlines

killercup committed:
Cow: 'static

killercup committed:
Cow: In the wild

killercup commented:
Thanks, @mbrubeck! Interesting trick to use a Cow in the benchmark to prevent the first alloc!

I fixed the typos, and added a few sections, to mention Cows with a static lifetime, `#[serde(borrow)]`, as well as your example.

killercup committed:
Cow: formatting

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+Making sure our data gets removed from memory so we don't use up too much memory and slow down the system is easy in Rust.
+Rust uses the ownership model of automatically `drop`ping resources when they go out of scope,
+so it doesn't need to periodically run a garbage collector to free memory.
+You can still waste memory, of course, but allocating too much of it manually,
```
`but allocating` => `by allocating`

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+
+[crossbeam-deque]: https://crates.io/crates/crossbeam-deque
+
+Making sure our data gets removed from memory so we don't use up too much memory and slow down the system is easy in Rust.
```
This sentence is a little long to read, maybe separate it into two or trim it?

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+(Carrying the length with us is a convention in Rust.)
+
+But what about a more complicated function?
+Let's image we want to replace some characters in a string.
```
`image` => `imagine`

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+  (to Cow own the data -- it's valid until the Cow goes out of scope),
+  but in case the Cow contains `Borrowed` data,
+  this lifetime is a restriction set by the data we refer to.
+  We cannot have a Cow to refers to already freed memory,
```
I think `Cow to refers` => `Cow that refers`

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+  but in case the Cow contains `Borrowed` data,
+  this lifetime is a restriction set by the data we refer to.
+  We cannot have a Cow to refers to already freed memory,
+  and rustc will us when that is possible by mentioning that the Cow outlives it's `'a`.
```
I'm not sure what you're trying to say at `rustc will us when`

alexcameron89 replied:
Also `outlive it's`  => `outlive its`

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+- `ToOwned` is a trait that defines a method to convert borrowed data into owned data
+  (by cloning it and giving us ownership of the new allocation, most likely).
+  The type we receive from this method is an associated type on the trait,
+  and it's name is `Owned` (yep, the same name as the Cow variant, sorry).
```
Another `it's` to `its`

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+If you want to replace the type of `bar` in
+`struct Foo { bar: String }`
+with a `Cow`,
+you'll have to specify the lifetime of the referece the `Cow` can include:
```
`reference` is misspelled

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+and everytime you take or return `Foo` you might just need to annotate it.
+
+One easy way around this is to use `'static'`:
+You can omit the lifetime annotation on your strcut,
```
`struct` is misspelled

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+
+One example for improving program performance by using a Cow is
+[this part][regex-redux-cow] of the Regex Redux micro-benchmark.
+The trick is to store a reference to the data a first
```
I think your meant to say `at first`

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+[regex-redux-cow]: https://github.com/TeXitoi/benchmarksgame-rs/blob/f78f21bffc68cb42dd9311694913ea798535e674/src/regex_redux.rs#L72-L79
+
+A great example for how you can use the super powers of Cows
+in you own structs
```
`you own` => `your own`

alexcameron89 commented on _drafts/2018-05-30-secret-life-of-cows.md:
```diff
@@ -0,0 +1,201 @@
+A great example for how you can use the super powers of Cows
+in you own structs
+to refer to input data instead of coyping it over
+is serde's `#[serde(borrow)]` attribute.
```
Nitpick: capitalize Serde

alexcameron89 reviewed (commented):
Thank you for letting me review, and for writing this blog! I haven't used 🐄s yet, so it was helpful for me to see how it can be used.

killercup committed:
Cow: Fix a truckload of typos and improve wording

killercup commented:
Thanks so much @alexcameron89!

killercup committed:
Cow: Fix more typos

This time thanks to vscode-spell-checker

killercup committed:
Cow: Add toc

killercup committed:
Cow: Path::to_string_lossy example

killercup committed:
Cow: Rewrite intro

Thanks, Robert!

killercup committed:
Cow: Headlines

killercup committed:
Cow: Stack, Head, References

killercup committed:
Cow: Add a ton of links

killercup committed:
Cow: Update date

killercup committed:
Cow: Links

killercup commented:
Alright, let's merge (and thus publish) this. If you have any more comments feel free to post them here or as issues.
